### PR TITLE
Feature/bigger custom penalty text field

### PR DESCRIPTION
**Pull request by Shirkanesi**
As requested in #131 I started implementing a bigger text-box for custom penalties.

To not surprise the user, there is now a config-option enabling this feature.
Also the way `ENTER` and `TAB` is handled in multiline-text is a bit hacky, since I tried to still allow pressing `ENTER` to create the penalty. This is done through listening to key-presses in the text-box and (in case of `ENTER` or `TAB`) replacing the typed caracter and executing the proper action.
If you want to create a newline or tab character, you need to type `SHIFT + (ENTER | TAB)`

However, there are still things to test:
(would be nice, if someone with this hard-/software could test this)
- [ ] Test the scaling of the multiline-textbox on higher screen resolutions
  - [x] 2736 x 1824
  - [ ] 1440p
  - [x] 4K
- [x] Test the emulation of `ENTER` and `TAB` presses in the multiline-box on different operating systems (hence this might depend on the line-separator of the OS).
Tested so far: 
  - [x] Windows 11 
  - [x] Windows 10
  - [x] Linux (Mint or Ubuntu)
  - [x] Mac OS

### CODE REVIEW

**Comment by Shirkanesi** [comment]
New layout:
grafik
Old layout:
grafik

**Comment by dfuchss** [comment]
Looks nice :)

**Comment by Shirkanesi** [comment]
Tested scaling on my Surface pro 7 (`2736 x 1824`); seems to scale well. I think it's ready for review.

**Comment by dfuchss** [comment]
I'll have a look at it tomorrow or on tuesday.

File: .gitignore [hunk 1]
```diff
@@ -39,3 +39,10 @@ buildNumber.properties
 .mvn/wrapper/maven-wrapper.jar
 .polyglot.*
 .settings
+jvm/.project
```
Comment (dfuchss) [inline .gitignore:42]: Projects & Classpath files of Eclipse projects shall not be in the .gitignore :)

File: bundles/edu.kit.kastel.sdq.eclipse.grading.api/src/edu/kit/kastel/sdq/eclipse/grading/api/PreferenceConstants.java [hunk 2]
```diff
@@ -11,6 +11,7 @@
 	public static final String ARTEMIS_URL = "artemisUrl";
 	public static final String ARTEMIS_USER = "artemisUser";
 	public static final String ARTEMIS_PASSWORD = "artemisPassword";
+	public static final String PREFFERES_LARGE_PENALTY_TEXT_PATH = "userPreferresLargePenaltyText";
```
Comment (dfuchss) [inline bundles/edu.kit.kastel.sdq.eclipse.grading.api/src/edu/kit/kastel/sdq/eclipse/grading/api/PreferenceConstants.java:14]: Typo: prefers

File: bundles/edu.kit.kastel.sdq.eclipse.grading.view/src/edu/kit/kastel/eclipse/grading/view/assessment/CustomButtonDialog.java [hunk 3]
```diff
@@ -51,22 +55,40 @@ protected void configureShell(Shell newShell) {
...
+		final GridData data = new GridData(SWT.FILL, SWT.CENTER, true, false);
+		
+		GridData customMessageInputFieldData;
+		if (userWantsBigWindow) {
+			this.customMessageInputField = new Text(comp, SWT.MULTI | SWT.BORDER | SWT.V_SCROLL  |SWT.H_SCROLL);
+			customMessageInputFieldData = new GridData(GridData.FILL_BOTH);
+			
+			// Calculating height and width based on the lineHeight (theoretically) ensures proper scaling across screen-sizes.
+			// However lacking of a 4K-screen this has not been tested entirely.
+			customMessageInputFieldData.minimumHeight = this.customMessageInputField.getLineHeight() * 8;
```
Comment (dfuchss) [inline bundles/edu.kit.kastel.sdq.eclipse.grading.view/src/edu/kit/kastel/eclipse/grading/view/assessment/CustomButtonDialog.java:76]: You may create constants for that.

Comment (dfuchss) [reply bundles/edu.kit.kastel.sdq.eclipse.grading.view/src/edu/kit/kastel/eclipse/grading/view/assessment/CustomButtonDialog.java:76]: also ff.

File: bundles/edu.kit.kastel.sdq.eclipse.grading.view/src/edu/kit/kastel/eclipse/grading/view/assessment/CustomButtonDialog.java [hunk 4]
```diff
@@ -87,4 +109,64 @@ protected void okPressed() {
...
+			if (e.keyCode == SWT.SHIFT) {
+				this.isShiftPressed = false;
+			}
+			
+			if (!this.isShiftPressed) {
+				if (e.keyCode == SWT.TAB || isReturnCharacter(e.keyCode)) {
+					int insertedLength;
+					if (isReturnCharacter(e.keyCode)) {
+						// Required due to Windows using \r\n, UNIX-like systems just \n
+						insertedLength = System.lineSeparator().length();
```
Comment (dfuchss) [inline bundles/edu.kit.kastel.sdq.eclipse.grading.view/src/edu/kit/kastel/eclipse/grading/view/assessment/CustomButtonDialog.java:138]: What about a constant?

File: bundles/edu.kit.kastel.sdq.eclipse.grading.view/src/edu/kit/kastel/eclipse/grading/view/preferences/ArtemisGradingPreferencesPage.java [hunk 5]
```diff
@@ -27,6 +27,7 @@
 	private BooleanFieldEditor isRelativeConfigPath;
 	private StringFieldEditor relativeConfigPath;
 	private FileFieldEditor absoluteConfigPath;
+	private BooleanFieldEditor userPreferresLargePenaltyText;
```
Comment (dfuchss) [inline bundles/edu.kit.kastel.sdq.eclipse.grading.view/src/edu/kit/kastel/eclipse/grading/view/preferences/ArtemisGradingPreferencesPage.java:30]: Typo :)

**Review by dfuchss — CHANGES_REQUESTED** [review]
Overall code looks really good, just some remarks.

**Comment by Shirkanesi** [comment]
Did you test it on another screen-resolution? Did it scale well on your screen?

File: .gitignore [hunk 6]
```diff
@@ -39,3 +39,4 @@ buildNumber.properties
 .mvn/wrapper/maven-wrapper.jar
 .polyglot.*
 .settings
+
```
Comment (dfuchss) [inline .gitignore:42]: You could delete this modification as well :D

**Review by dfuchss — CHANGES_REQUESTED** [review]
Code looks good, currently I'm testing it.

**Comment by dfuchss** [comment]
Code works well. (I've created some random comment)
The only thing is that the multi line comments are not shown in Eclipse (only first line) but I think that may be a problem of eclipse
image

The presentation in Artemis is of. But maybe this should be adapted later on (not in this PR)
image

File: .gitignore [hunk 7]
```diff
@@ -36,6 +36,3 @@ dependency-reduced-pom.xml
 buildNumber.properties
 .mvn/timing.properties
 # https://github.com/takari/maven-wrapper#usage-without-binary-jar
-.mvn/wrapper/maven-wrapper.jar
```
Comment (dfuchss) [inline .gitignore:39]: Now you've deleted some ignores :D

Comment (dfuchss) [reply .gitignore:39]: See https://github.com/kit-sdq/programming-lecture-eclipse-artemis/pull/138/files

**Comment by Shirkanesi** [comment]
Yes, eclipse does some interesting things when displaying those comments (seems linke the new-lines are just hidden in the preview. However in the Properties of the marker they seem to work as expected).

Artemis does not realy use the multiline with \n, however <br> does work sometimes (but somethimes <br> is displayed directly in the text...). I think this needs some more investigation in the future.

**Comment by Shirkanesi** [comment]
If you want, we can include the <br>

File: .gitignore (hunk 7 above)
Comment (dfuchss) [reply .gitignore:39]: Sorry, you have to fix the last line, then I'll merge the PR 😄

**Comment by dfuchss** [comment]
> If you want, we can include the `<br> `

Maybe thats something for later improvements :)

File: .gitignore (hunk 7 above)
Comment (Shirkanesi) [reply .gitignore:39]: The current version is the same as in main? I think?

**Comment by dfuchss** [comment]
@Shirkanesi see comment above. If the last change at gitignore is reverted, I'll be very happy to merge this feature :)

File: .gitignore (hunk 7 above)
Comment (dfuchss) [reply .gitignore:39]: Newline is missing as far as I see :D

**Comment by dfuchss** [comment]
Nice work !

**Pull request merged**